### PR TITLE
Create introspection plugin

### PR DESCRIPTION
**Pull request by RanAbram**

### CODE REVIEW

File: src/access.lua [hunk 1]
```diff
@@ -68,13 +70,16 @@ function _M.execute(conf)
...
   local line, err = sock:receive("*l")
-
   if err then 
     ngx.log(ngx.ERR, name .. "failed to read response status from " .. host .. ":" .. tostring(port) .. ": ", err)
-    return
+    return kong_response.exit(500, "internal error")
   end
 
   local status_code = tonumber(string.match(line, "%s(%d%d%d)%s"))
+  if status_code > 299 then
```
Comment (yurinovo) [inline src/access.lua]: I think it should be > 399 ?

File: src/access.lua [hunk 2]
```diff
@@ -94,30 +99,30 @@ function _M.execute(conf)
...
   if not ok then
     ngx.log(ngx.ERR, name .. "failed to keepalive to " .. host .. ":" .. tostring(port) .. ": ", err)
-    return
+    return kong_response.exit(500, "failed to keepalive to host")
   end
 
-  if status_code > 299 then
-    if err then 
-      ngx.log(ngx.ERR, name .. "failed to read response from " .. host .. ":" .. tostring(port) .. ": ", err)
+  if status_code == 200 then
```
Comment (yurinovo) [inline src/access.lua]: it can also be 204.... I think you should check 200 <= status_code <= 399

File: src/access.lua [hunk 3]
```diff
@@ -94,30 +99,30 @@ function _M.execute(conf)
...
-    if conf.response == "table" then 
-      response_body = JSON:decode(string.match(body, "%b{}"))
+    if response_params['code'] == 200 then
+      for key, value in pairs(response_params) do
+        ngx_set_header("Introspection-" .. string.gsub(key, "_", "-"), value)
+      end
+      return
     else
-      response_body = string.match(body, "%b{}")
+      return kong_response.exit(response_params['code'], "failed to authenticate token")
```
Comment (yurinovo) [inline src/access.lua]: failed to authenticate token -> failed to authenticate utoken ?

**Review by yurinovo — CHANGES_REQUESTED** [review]

File: src/access.lua [hunk 4]
```diff
@@ -94,69 +112,66 @@ function _M.execute(conf)
...
+    if introspection_details.query then
+      introspection_uri_args = url_utils.parseQuery(introspection_details.query)
+    end
 
-    local url
-    if parsed_url.query then
-      url = parsed_url.path .. "?" .. parsed_url.query
-    else
-      url = parsed_url.path
+    local app_key = extract_param(request_uri_args, request_body, "app_key")
```
Comment (RanAbram) [inline src/access.lua]: Remove

File: src/access.lua [hunk 5]
```diff
@@ -34,6 +36,19 @@ local function parse_url(host_url)
   return parsed_url
 end
 
+local function extract_param(url_params, body_params, attribute_name)
+  local extracted_value = nil
+  if url_params then
```
Comment (RanAbram) [inline src/access.lua]: `url_params && url_params[attribute_name]`

File: src/access.lua [hunk 6]
```diff
@@ -34,6 +36,19 @@ local function parse_url(host_url)
...
+    extracted_value = url_params[attribute_name]
+  end
+
+  if not extracted_value and body_params then
```
Comment (RanAbram) [inline src/access.lua]: remove `not extracted_value`

File: src/access.lua [hunk 7]
```diff
@@ -34,6 +36,19 @@ local function parse_url(host_url)
...
+    extracted_value = JSON:decode(body_params)[attribute_name]
```
Comment (RanAbram) [inline src/access.lua]: Check failure points and use save checks (JSON:decode_with_...)

File: src/access.lua [hunk 8]
```diff
@@ -94,69 +112,66 @@ function _M.execute(conf)
...
-      response_body = JSON:decode(string.match(body, "%b{}"))
+    is_token_valid = response_params["valid_token"]
+    if is_token_valid then
+      for key, value in pairs(response_params) do
+        ngx_set_header("Introspection-" .. string.gsub(key, "_", "-"), value)
+      end
+      return
     else
-      response_body = string.match(body, "%b{}")
+      return kong_response.exit(404, "utoken not authenticated")
```
Comment (RanAbram) [inline src/access.lua]: 401

File: src/access.lua (hunk 2 above)
Comment (RanAbram) [reply src/access.lua]: Remove condition

File: src/access.lua [hunk 9]
```diff
@@ -94,69 +112,66 @@ function _M.execute(conf)
...
+      response_params[string.lower(key)] = value
     end
 
-    local response_body
-    if conf.response == "table" then 
-      response_body = JSON:decode(string.match(body, "%b{}"))
+    is_token_valid = response_params["valid_token"]
+    if is_token_valid then
+      for key, value in pairs(response_params) do
+        ngx_set_header("Introspection-" .. string.gsub(key, "_", "-"), value)
```
Comment (RanAbram) [inline src/access.lua]: Remove gsub

File: src/access.lua [hunk 10]
```diff
@@ -94,69 +112,66 @@ function _M.execute(conf)
...
+    end
 
-    local url
-    if parsed_url.query then
-      url = parsed_url.path .. "?" .. parsed_url.query
-    else
-      url = parsed_url.path
+    local app_key = extract_param(request_uri_args, request_body, "app_key")  
+    local utoken = extract_param(request_uri_args, request_body, "utoken")
+    if not utoken then
```
Comment (RanAbram) [inline src/access.lua:131]: Consider remove.
Check for `params[:token]`
Check under b2b

File: src/access.lua [hunk 11]
```diff
@@ -94,69 +112,66 @@ function _M.execute(conf)
...
+      utoken = extract_param(request_uri_args, request_body, "token")
     end
-    
-    local raw_json_headers = JSON:encode(headers)
-    local raw_json_body_data = JSON:encode(body_data)
 
-    local raw_json_uri_args
-    if next(uri_args) then 
-      raw_json_uri_args = JSON:encode(uri_args) 
+    introspection_uri_args["app_key"] = app_key
```
Comment (RanAbram) [inline src/access.lua]: Remove

File: src/access.lua (hunk 9 above)
Comment (RanAbram) [reply src/access.lua]: Change introspection `X-Introspection-`

File: src/access.lua (hunk 1 above)
Comment (RanAbram) [reply src/access.lua]: Done

File: src/access.lua (hunk 2 above)
Comment (RanAbram) [reply src/access.lua]: Done

File: src/access.lua (hunk 8 above)
Comment (RanAbram) [reply src/access.lua]: Done

File: src/access.lua (hunk 4 above)
Comment (RanAbram) [reply src/access.lua]: Done

File: src/access.lua [hunk 12]
```diff
@@ -52,111 +56,101 @@ function _M.execute(conf)
...
     ngx.log(ngx.ERR, name .. "failed to read response status from " .. host .. ":" .. tostring(port) .. ": ", err)
-    return
+    return kong_response.exit(500, "internal error")
   end
 
   local status_code = tonumber(string.match(line, "%s(%d%d%d)%s"))
+  if status_code > 399 then
+    local error_message = "internal error"
+    if status_code == 401 then
+      error_message = "utoken not authenticated"
```
Comment (yurinovo) [inline src/access.lua]: "unauthorized" I think is enough or "bad credentials"

File: src/access.lua [hunk 13]
```diff
@@ -52,111 +56,101 @@ function _M.execute(conf)
...
+
+    local decoded_body, err = JSON:decode(line)
+    if err then
+      ngx.log(ngx.ERR, name .. "failed to parse body " .. host .. ":" .. tostring(port) .. ": ", err)
+      return kong_response.exit(500, "internal error")
+    end
+    if type(decoded_body) == "table" then
+      body = decoded_body
+    end
+  until not decoded_body
```
Comment (yurinovo) [inline src/access.lua]: not sure I got it,
each loop you read one line of the body
how do we know that the entire body is one line? can't it be more than one line?

**Review by yurinovo — CHANGES_REQUESTED** [review]

**Pull request merged**